provider/vsphere/vm_template.go

```diff
@@ -0,0 +1,206 @@
+	}
+
+	logger.Debugf("Looking for already imported templates for arches %q", arches)
+	// Attempt to find a previously imported instance template
+	importedTemplate, arch, err := v.getImportedTemplate(ctx, series, arches)
+	if err == nil {
+		logger.Debugf("Found already imported template for series %s", series)
+		return importedTemplate, arch, nil
+	}
+
+	// Ignore not found errors. It means we have not imported a template yet.
+	if !errors.IsNotFound(err) {
+		return nil, "", errors.Trace(err)
+	}
+	logger.Debugf("Downloading and importing template from simplestreams")
+	// Last resort, download and import a template.
+	return v.downloadAndImportTemplate(ctx, series, arches)
+}
+
+// findTemplate uses the imageMetadata provided to find a local template.
+// The imageMetadata parameter holds a list of already filtered templates,
+// that should match the series that was requested.
+func (v *vmTemplateManager) findTemplate(ctx context.Context) (*object.VirtualMachine, string, error) {
+	if len(v.imageMetadata) == 0 {
+		return nil, "", errors.NotFoundf("no image metadata was supplied")
+	}
+
+	for _, img := range v.imageMetadata {
+		vms, err := v.env.client.ListVMTemplates(ctx, img.Id)
+		if err != nil {
+			return nil, "", errors.Trace(err)
+		}
+		switch len(vms) {
+		case 1:
+			// trust that due diligence was made when generating simplestreams
+			// and the img.Arch, reflects the architecture of the OS running insude
+			// the VM generated from the found template.
+			return vms[0], img.Arch, nil
+		default:
+			continue
+		}
+	}
+	return nil, "", errors.NotFoundf("could not find a suitable template")
+}
+
+func (v *vmTemplateManager) seriesTemplateFolder(series string) string {
+	templatesPath := v.env.controllerTemplatesFolder(v.controllerUUID)
+	return path.Join(templatesPath, series)
+}
+
+func (v *vmTemplateManager) getVMArch(ctx context.Context, vmObj *object.VirtualMachine) (string, error) {
+	vmMo, err := v.env.client.VirtualMachineObjectToManagedObject(ctx, vmObj)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	for _, item := range vmMo.Config.ExtraConfig {
+		value := item.GetOptionValue()
+		if value.Key == vsphereclient.ArchTag {
+			arch, ok := value.Value.(string)
+			if !ok {
+				return "", errors.Errorf("invalid arch tag for VM")
+			}
+			return arch, nil
+		}
+	}
+	return "", errors.NotFoundf("arch tag not found for VM")
+}
+
+func (v *vmTemplateManager) isValidArch(arch string, desiredArches []string) bool {
+	for _, item := range desiredArches {
+		if item == arch {
+			return true
+		}
+	}
+	return false
+}
+
+func (v *vmTemplateManager) getImportedTemplate(ctx context.Context, series string, arches []string) (*object.VirtualMachine, string, error) {
+	seriesTemplatesFolder := v.seriesTemplateFolder(series)
+	seriesTemplates, err := v.env.client.ListVMTemplates(ctx, path.Join(seriesTemplatesFolder, "*"))
+	if err != nil {
+		logger.Errorf("failed to fetch templates: %v", err)
```

Comment:
This causes misleading error messages during bootstrap, when the folder for the bootstrapping controller hasn't been created yet.
```
...
Launching controller instance(s) on boston-vsphere/Boston...
ERROR failed to fetch templates: path not found: juju-ci-root/Juju Controller (baf1f6a2-f6d7-44f1-8916-7165113f7f8f)/templates/focal/* not found
 - juju-dc6839-0 (arch=amd64 mem=3.5G) f81b245c98d18fec1564e6b374c346f1ef14394c9e836ee77e1e58754242a4f9"
Installing Juju agent on bootstrap instance
...
```